plugins/chat/app/helpers/chat/with_service_helper.rb

```diff
@@ -19,11 +20,19 @@ def with_service(service, default_actions: true, **dependencies, &block)
     end
 
     def run_service(service, dependencies)
-      @_result = service.call(params.to_unsafe_h.merge(guardian: guardian, **dependencies))
+      @_result =
```

Comment:
LMK your thoughts on the changes here @Flink , I did them so we can also use `with_service` in specs, e.g. see https://github.com/discourse/discourse/pull/22144/files#diff-fffd7577aaa7e572a63261269cd5b0a1ada0b80b744799412d110a717f50c376R21

Reply:
@Flink yes, @jjaffeux and I have discussed this and a few weeks back we changed the `:chat_message` fabricator to use the `Chat::MessageCreator` service, since we want things to reflect reality more especially for system specs. We will need to do something like this if we ever want to call services outside of controllers too...

Joffrey what are your thoughts on this? I don't have a super strong opinion on it so happy to defer to you/Loic

Reply:
Well, I guess it’s always about trade-offs :sweat_smile: Indeed, by using the service, we’re sure to have the exact same output than in production. But as you said @jjaffeux, testing the side effects should happen in the service specs and not outside. In other specs, if we need to create a DM channel, then it’s probably just a setup step and I’m pretty sure most of the time we won’t be interested by anything else than just having a channel properly created.

Using fabricators only, we have to replicate some of the logic we have in the service and have to keep it up-to-date. In the end, it depends on what we need. I think using fabricators should be enough but I could be wrong :slightly_smiling_face: 

If we wanted to use the service directly, actually we could do this by using fabricators too as they’re able to call any object.

Finally, we don’t really need `with_service` in specs, we can simply call the service directly and provide proper arguments as we control everything here. We shouldn’t have to handle the service failures as we want a specific state for our tests, so if it’s failing then it means our setup is not a good one.

Reply:
> If we wanted to use the service directly, actually we could do this by using fabricators too as they’re able to call any object.

Yeah this is what we are already doing with the `:chat_message` fabricator.

> We shouldn’t have to handle the service failures as we want a specific state for our tests, so if it’s failing then it means our setup is not a good one.

I suppose this is a good point, I mainly wanted a nice way to automatically get a spec failure + be able to inspect it when a called service fails rather than having to do it manually in each test.

> Maybe using it for Chat::Message was not a good idea too, and the fact that we thought it was is probably showing us shortcomings...

It is hard to tell...the side-effects are really necessary for system specs IMO for them to be realistic, but yeah probably overkill for normal unit tests like this :thinking: Maybe we could have a version of the fabricator which uses the service class and a version without and use the appropriate one based on the test? E.g. `Fabricate(:chat_message_with_service)` for system tests and `Fabricate(:chat_message)` for unit specs? Not sure...

@jjaffeux @Flink I have no idea how github notifies based on comments so just mention you both again :)

Reply:
Indeed, if we need message bus events to be sent or similar things, it’s better to call the service directly. This idea of having different fabricators also crossed my mind, so maybe that’s something to explore?

Reply:
Yep Im not sure yet too... so far having fabricators without side effects have caused me way more issues than having fabricators being the actual service call, that's all I can tell.

Reply:
Indeed, when you need all the “real” effects like in system specs it’s very convenient. But in all the other specs, I think it’s overkill (and slower, etc.)

Reply:
So, are we keeping this in the end? I’m pretty convinced now that having `with_service` in specs is not necessary, but we can just call the service in fabricators (this way it’s relatively consistent with how we setup things in specs).

Reply:
TBH I don't want to have to mess around with changing the fabricator for DM channel at the moment, I want to get this merged, so I will just revert the `with_service` changes in this PR and call the service directly in the specs